opt: panic while executing 1 statements: EXPLAIN (DISTSQL) SELECT _ FROM _._._ W...

https://sentry.io/cockroach-labs/cockroachdb/issues/694086358/

```
*log.safeError: conn_executor.go:668: panic while executing 1 statements: EXPLAIN (DISTSQL) SELECT _ FROM _._._ WHERE (SELECT count(_) >= _) GROUP BY _: caused by <redacted>
  File "github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go", line 414, in func1
  File "github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go", line 109, in indexedVar
  File "github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go", line 101, in buildVariable
  File "github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go", line 87, in buildScalar
  File "github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go", line 177, in buildComparison
...
(21 additional frame(s) were not displayed)

conn_executor.go:668: panic while executing 1 statements: EXPLAIN (DISTSQL) SELECT _ FROM _._._ WHERE (SELECT count(_) >= _) GROUP BY _: caused by <redacted>
```

Rebecca, I have a repro for a crash that looks like this query, though it fails in a different place:
```
CREATE TABLE a (k INT PRIMARY KEY, i INT, f FLOAT NOT NULL, s STRING, j JSON)
SELECT s FROM a WHERE (SELECT COUNT(i) >= 100) GROUP BY s
```
Here's the crash with this query:
```
SELECT s FROM a WHERE (SELECT COUNT(i) >= 100) GROUP BY s
panic: top-level relational expression cannot have outer columns: (7) [recovered]
	panic: top-level relational expression cannot have outer columns: (7) [recovered]
	panic: top-level relational expression cannot have outer columns: (7)

goroutine 16 [running]:
testing.tRunner.func1(0xc4201e9950)
	/usr/local/go/src/testing/testing.go:742 +0x29d
panic(0x4c76600, 0xc420255620)
	/usr/local/go/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/testutils/datadriven.RunTest.func1.1(0xc42047a720)
	/Users/andy/go/src/github.com/cockroachdb/cockroach/pkg/testutils/datadriven/datadriven.go:75 +0x117
panic(0x4c76600, 0xc420255620)
	/usr/local/go/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).Optimize(0xc42024c000, 0x0, 0xc42024c000)
	/Users/andy/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:182 +0x200
```
I believe it's due to aggregations + correlated subqueries.

--- a/pkg/sql/opt/optbuilder/scalar.go
+++ b/pkg/sql/opt/optbuilder/scalar.go
@@ -579,9 +579,13 @@ func (b *Builder) buildRangeCond(
 
 // checkSubqueryOuterCols uses the subquery outer columns to update the given
 // set of column references and the set of outer columns for any enclosing
-// subuqery. If this is a grouping context, it checks that any outer columns
-// from the given subquery that reference inScope are either aggregate or
-// grouping columns in inScope.
+// subuqery. It also performs the following checks:
+//   1. If aggregates are not allowed in the current context (e.g., if we
+//      are building the WHERE clause), it checks that the subquery does not
+//      reference any aggregates from this scope.
+//   2. If this is a grouping context, it checks that any outer columns from
+//      the given subquery that reference inScope are either aggregate or
+//      grouping columns in inScope.
 func (b *Builder) checkSubqueryOuterCols(
 	subqueryOuterCols opt.ColSet, inGroupingContext bool, inScope *scope, colRefs *opt.ColSet,
 ) {
@@ -602,6 +606,19 @@ func (b *Builder) checkSubqueryOuterCols(
 		b.subquery.outerCols.UnionWith(subqueryOuterCols.Difference(inScopeCols))
 	}
 
+	// Check 1 (see function comment).
+	if b.semaCtx.Properties.IsSet(tree.RejectAggregates) && inScope.groupby.aggOutScope != nil {
+		aggCols := inScope.groupby.aggOutScope.getAggregateCols()
+		for i := range aggCols {
+			if subqueryOuterCols.Contains(int(aggCols[i].id)) {
+				panic(builderError{
+					tree.NewInvalidFunctionUsageError(tree.AggregateClass, inScope.context),
+				})
+			}
+		}
+	}
+
+	// Check 2 (see function comment).
 	if inGroupingContext {
 		subqueryOuterCols.IntersectionWith(inScopeCols)
 		if !subqueryOuterCols.Empty() &&

--- a/pkg/sql/opt/optbuilder/testdata/subquery
+++ b/pkg/sql/opt/optbuilder/testdata/subquery
@@ -2466,3 +2466,75 @@ project
                      │              └── array-agg [type=int[]]
                      │                   └── variable: x [type=int]
                      └── array: [type=int[]]
+
+# Regression test for #30424. Aggregate function in the WHERE subquery
+# is aggregated in the outer scope, so it's not allowed.
+build
+SELECT s FROM a WHERE (SELECT count(i) >= 100) GROUP BY s
+----
+error (42803): aggregate functions are not allowed in WHERE
+
+# Aggregate function in the WHERE subquery is aggregated in the subquery scope,
+# so it's allowed.
+build
+SELECT s FROM a WHERE (SELECT count(i) >= 100 FROM a) GROUP BY s
+----
+group-by
+ ├── columns: s:4(string)
+ ├── grouping columns: a.s:4(string)
+ └── project
+      ├── columns: a.s:4(string)
+      └── select
+           ├── columns: a.k:1(int!null) a.i:2(int) a.f:3(float) a.s:4(string) a.j:5(jsonb)
+           ├── scan a
+           │    └── columns: a.k:1(int!null) a.i:2(int) a.f:3(float) a.s:4(string) a.j:5(jsonb)
+           └── filters [type=bool]
+                └── subquery [type=bool]
+                     └── max1-row
+                          ├── columns: "?column?":12(bool)
+                          └── project
+                               ├── columns: "?column?":12(bool)
+                               ├── scalar-group-by
+                               │    ├── columns: count:11(int)
+                               │    ├── project
+                               │    │    ├── columns: a.i:7(int)
+                               │    │    └── scan a
+                               │    │         └── columns: a.k:6(int!null) a.i:7(int) a.f:8(float) a.s:9(string) a.j:10(jsonb)
+                               │    └── aggregations
+                               │         └── count [type=int]
+                               │              └── variable: a.i [type=int]
+                               └── projections
+                                    └── ge [type=bool]
+                                         ├── variable: count [type=int]
+                                         └── const: 100 [type=int]
+
+exec-ddl
+CREATE TABLE xyzs (x INT PRIMARY KEY, y INT, z FLOAT NOT NULL, s STRING, UNIQUE (s DESC, z));
+----
+TABLE xyzs
+ ├── x int not null
+ ├── y int
+ ├── z float not null
+ ├── s string
+ ├── INDEX primary
+ │    └── x int not null
+ └── INDEX secondary
+      ├── s string desc
+      ├── z float not null
+      └── x int not null (storing)
+
+exec-ddl
+CREATE TABLE kuv (k INT PRIMARY KEY, u FLOAT, v STRING);
+----
+TABLE kuv
+ ├── k int not null
+ ├── u float
+ ├── v string
+ └── INDEX primary
+      └── k int not null
+
+# Regression test for #29114.
+build
+SELECT * FROM xyzs WHERE (SELECT sum(x) FROM (SELECT u FROM kuv) GROUP BY u) > 100;
+----
+error (42803): aggregate functions are not allowed in WHERE

--- a/pkg/sql/sem/tree/type_check.go
+++ b/pkg/sql/sem/tree/type_check.go
@@ -102,6 +102,11 @@ func (s *SemaProperties) Require(context string, rejectFlags SemaRejectFlags) {
 	s.Derived.Clear()
 }
 
+// IsSet checks if the given rejectFlag is set as a required property.
+func (s *SemaProperties) IsSet(rejectFlags SemaRejectFlags) bool {
+	return s.required.rejectFlags&rejectFlags != 0
+}
+
 // Restore restores a copy of a SemaProperties. Use with:
 // defer semaCtx.Properties.Restore(semaCtx.Properties)
 func (s *SemaProperties) Restore(orig SemaProperties) {

